src/gtp_redirector.erl

```diff
@@ -0,0 +1,334 @@
+%% Copyright 2017, Travelping GmbH <info@travelping.com>
+
+%% This program is free software; you can redistribute it and/or
+%% modify it under the terms of the GNU General Public License
+%% as published by the Free Software Foundation; either version
+%% 2 of the License, or (at your option) any later version.
+
+-module(gtp_redirector).
+
+-export([init/2, 
+         validate_options/1, 
+         keep_alive/2, 
+         timeout_requests/1, 
+         handle_message/6, 
+         echo_response/6]).
+
+-type nodes_map() :: #{binary() => {v1 | v2, inet:ip_address()}}.
+-type condition() :: {sgsn_ip, inet:ip_address()} |
+                     {version, list(v1 | v2)} |
+                     {imsi, binary()}.
+-type rule():: #{conditions => list(condition()),
+                 nodes => list(binary())}.
+
+-record(redirector, {
+          socket     = nil    :: nil | gen_socket:socket(),
+
+          % lists of all nodes, including bad
+          nodes      = #{}    :: nodes_map(),
+
+          % here we have nodes which are not answered for Keep-Alive echo request
+          % no requests should be transfered to such nodes
+          bad_nodes  = #{}    :: nodes_map(),
+          lb_type    = random :: random | round_robin, % for now only random is supported
+          ka_timeout = 60000  :: non_neg_integer(), % Keep-Alive timeout
+          ka_timer   = nil    :: reference(),
+
+          rules      = []  :: list(rule()),
+
+          % cached nodes for requests, because retransmission should happen for the same node
+          requests,         % :: ergw_cache(),
+          rt_timeout = 10000  :: non_neg_integer(), % retransmission timeout
+
+          % timestamps for responses which used by Keep-Alive mechanism
+          responses  = maps:new() :: map()
+         }).
+
+-type redirector() :: #redirector{}.
+
+-export_type([redirector/0]).
+
+-include_lib("gtplib/include/gtp_packet.hrl").
+
+-include("include/ergw.hrl").
+
+-define(T3, 10 * 1000).
+
+init(IP, #{redirector := #{keep_alive_timeout := KATimeout,
+                           retransmit_timeout := RTTimeout,
+                           lb_type := LBType,
+                           rules := Rules,
+                           nodes := Nodes} = _Opts}) -> 
+    {ok, Socket} = gen_socket:socket(gtp_socket:family(IP), raw, udp),
```

Comment:
this should use the same logic as gtp_socket to create and bind socket and handle socket options.